qiskit/aqua/algorithms/phase_estimators/phase_estimator.py

```diff
@@ -0,0 +1,199 @@
+    recorded. The results are returned as an instance of
+    qiskit.algorithms.phase_estimator_result.PhaseEstimatorResult.
+
+    If the input state is an eigenstate of the unitary, then in the ideal case, all probability is
+    concentrated on the bitstring corresponding to the eigenvalue of the input state. If the input
+    state is a superposition of eigenstates, then each bitstring is measured with a probability
+    corresponding to its weight in the superposition. In addition, if the phase is not representable
+    exactly by the phase-reading register, the probability will be spread across bitstrings, with an
+    amplitude that decreases with distance from the bitstring most closely approximating the phase.
+
+    **Reference:**
+
+    [1]: Michael A. Nielsen and Isaac L. Chuang. 2011.
+         Quantum Computation and Quantum Information: 10th Anniversary Edition (10th ed.).
+         Cambridge University Press, New York, NY, USA.
+
+    """
+
+    def __init__(self,
+                 num_evaluation_qubits: int,
+                 unitary: Optional[QuantumCircuit] = None,
+                 state_preparation: Optional[QuantumCircuit] = None,
+                 pe_circuit: Optional[QuantumCircuit] = None,
+                 num_unitary_qubits: Optional[int] = None,
+                 quantum_instance: Optional[Union[QuantumInstance, BaseBackend]] = None) -> None:
+
+        """Args:
+            num_evaluation_qubits: The number of qubits used in estimating the phase. The phase will
+                                   be estimated as a binary string with this many bits.
+            unitary: The circuit representing the unitary operator whose eigenvalues (via phase)
+                     will be measured. Exactly one of `pe_circuit` and `unitary` must be passed.
+            state_preparation: The circuit that prepares the state whose eigenphase will be
+                                 measured.  If this parameter is omitted, no preparation circuit
+                                 will be run and input state will be the all-zero state in the
+                                 computational basis.
+            pe_circuit: The phase estimation circuit.
+            num_unitary_qubits: Must agree with the number of qubits in the unitary in `pe_circuit`
+                                if `pe_circuit` is passed. This parameter will be set from `unitary`
+                                if `unitary` is passed.
+            quantum_instance: The quantum instance on which the circuit will be run.
+
+           Raises:
+               ValueError: unless only one of `unitary` and `pe_circuit` is `None`.
+                           `num_unitary_qubits` disagrees with size of `unitary`.
+        """
+
+        # Determine if user passed a unitary, or the entire QPE circuit with the unitary
+        # already embedded, and set properties.
+        if unitary is None:
```

Comment:
This seems to check if they are both None, i.e. we need either unitary or pe_circuit to be passed. There is no check if both are passed but if so it looks like unitary wins and pe_circuit as supplied is ignored. The ValueError implies that a maybe I would have expected that to have been raised.